### PR TITLE
[hw,mbx,dif] Sync MBX DIF from integrated_dev to Master 

### PR DESCRIPTION
**Pull request by Razer6**
This PR syncs the MBX DIF from `integrated_dev` to `master`. 

This PR depends on https://github.com/lowRISC/opentitan/pull/24437

### CODE REVIEW

File: sw/device/lib/dif/dif_mbx.h [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+  // Fields, if necessary.
+  uint32_t imbx_base_addr;
+  uint32_t imbx_limit_addr;
+  uint32_t ombx_base_addr;
+  uint32_t ombx_limit_addr;
+} dif_mbx_range_config_t;
+
+/**
+ * A DOE transaction is allowed to have at maximum 1024 double words (32-bit
+ * each)
```
Comment (engdoreis) [inline sw/device/lib/dif/dif_mbx.h:34]: For consistency.
```suggestion
 * each).
```

File: sw/device/lib/dif/dif_mbx.h [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+ * @param[out] doe_intr_data Mailbox interrupt value.
+ * @return The result of the operation.
+ */
+OT_WARN_UNUSED_RESULT
+dif_result_t dif_mbx_ipi_configuration_get(const dif_mbx_t *mbx,
+                                           uint32_t *doe_intr_addr,
+                                           uint32_t *doe_intr_data);
+
+/**
+ * Host reads the DoE Mailbox request from internal SRAM
```
Comment (engdoreis) [inline sw/device/lib/dif/dif_mbx.h:92]:
```suggestion
 * Host reads the DoE Mailbox request from internal SRAM.
```

File: sw/device/lib/dif/dif_mbx.h [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+ * @param mbx A DOE Mailbox handle.
+ * @param[out] request DOE object read from the internal SRAM.
+ * @return The result of the operation.
+ */
+OT_WARN_UNUSED_RESULT
+dif_result_t dif_mbx_process_request(const dif_mbx_t *mbx,
+                                     dif_mbx_transaction_t *request);
+
+/**
+ * Host writes the DoE Mailbox response to the internal SRAM
```
Comment (engdoreis) [inline sw/device/lib/dif/dif_mbx.h:103]:
```suggestion
 * Host writes the DoE Mailbox response to the internal SRAM.
```

File: sw/device/lib/dif/dif_mbx.c [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+}
+
+dif_result_t dif_mbx_is_busy(const dif_mbx_t *mbx, bool *is_busy) {
+  if (mbx == NULL || is_busy == NULL) {
+    return kDifBadArg;
+  }
+
+  *is_busy = bitfield_bit32_read(
+      mmio_region_read32(mbx->base_addr, MBX_STATUS_REG_OFFSET),
+      MBX_STATUS_BUSY_BIT);
```
Comment (engdoreis) [inline sw/device/lib/dif/dif_mbx.c:71]: I would avoid the implicit cast from `int` to` bool`.
```suggestion
  *is_busy = (bitfield_bit32_read(
      mmio_region_read32(mbx->base_addr, MBX_STATUS_REG_OFFSET),
      MBX_STATUS_BUSY_BIT) == 1);
```

File: sw/device/lib/dif/dif_mbx.md [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+This checklist is for [Development Stage](/doc/project_governance/development_stages.md) transitions for the [Mailbox DIF](/hw/ip/mbx/README.md).
+All checklist items refer to the content in the [Checklist](/doc/project_governance/checklist/README.md).
+
+<h2>DIF Checklist</h2>
+
+<h3>S1</h3>
+
+Type           | Item                   | Resolution  | Note/Collaterals
+---------------|------------------------|-------------|------------------
+Implementation | [DIF_EXISTS][]         | Not Started |
```
Comment (engdoreis) [inline sw/device/lib/dif/dif_mbx.md:22]: Can you please update this checklist?

**Review by engdoreis — APPROVED** [review]
LGTM.

**Comment by Razer6** [comment]
@engdoreis Thanks for the review. I updated the comments.

File: sw/device/lib/dif/dif_mbx.h [hunk 6]
```diff
@@ -5,6 +5,54 @@
...
 
+#ifdef __cplusplus
+extern "C" {
+#endif  // __cplusplus
+
+/**
+ * Inbound and outbound range for DOE Mailbox.
+ */
+typedef struct dif_mbx_range_config {
+  // Fields, if necessary.
```
Comment (pamaury) [inline sw/device/lib/dif/dif_mbx.h:25]: I do not understand this commit

File: sw/device/lib/dif/dif_mbx.c [hunk 7]
```diff
@@ -57,3 +58,56 @@ dif_result_t dif_mbx_range_get(const dif_mbx_t *mbx,
...
+    return kDifBadArg;
+  }
+
+  *is_busy = (bitfield_bit32_read(
+                  mmio_region_read32(mbx->base_addr, MBX_STATUS_REG_OFFSET),
+                  MBX_STATUS_BUSY_BIT) == 1);
+  return kDifOk;
+}
+
+dif_result_t dif_mbx_process_request(const dif_mbx_t *mbx,
```
Comment (pamaury) [inline sw/device/lib/dif/dif_mbx.c:90]: This function is a bit suspicious: it reads data from the mailbox into `request->data_words` but there seems to be no limit to how much data can be read (except for the mailbox limit), and it does not even take as input the maximum number of words that can be written.
I would suggest that this function uses `request->nr_dwords` as the maximum size, and updates it with the number of words written. Possibly it should indicate as well when it has reached the maximum but there was still more to read?

**Review by pamaury — CHANGES_REQUESTED** [review]
This looks good except for the function reading the mailbox where I think some more safety should be added (see comment in code)

File: sw/device/lib/dif/dif_mbx.h (hunk 6 above)
Comment (Razer6) [reply sw/device/lib/dif/dif_mbx.h:25]: Dropped that.

File: sw/device/lib/dif/dif_mbx.c (hunk 7 above)
Comment (Razer6) [reply sw/device/lib/dif/dif_mbx.c:90]: Agree. In the last commit I changed the function to limit the number words read by `request->nr_dwords`. If there is more to read, I'm returning a different return code.

File: sw/device/lib/dif/autogen/dif_mbx_autogen.c [hunk 8]
```diff
@@ -0,0 +1,263 @@
+// Copyright lowRISC contributors (OpenTitan project).
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+// THIS FILE HAS BEEN GENERATED, DO NOT EDIT MANUALLY. COMMAND:
+// util/make_new_dif.py --mode=regen --only=autogen
+
+#include "sw/ip/mbx/dif/autogen/dif_mbx_autogen.h"
```
Comment (pamaury) [inline sw/device/lib/dif/autogen/dif_mbx_autogen.c:8]: I think this was changed by mistake? This uses the path in integrated_dev

Comment (Razer6) [reply sw/device/lib/dif/autogen/dif_mbx_autogen.c:8]: Something went south here during the pull/rebase (2 machines). I'll take a look.

**Comment by Razer6** [comment]
I updated the documentation of that function.

**Pull request merged**